guacamole/src/main/frontend/src/app/import/controllers/importConnectionsController.js

```diff
@@ -0,0 +1,760 @@
+        !($scope.fileReader || $scope.dataReady);
+
+    /**
+     * Handle a provided File upload, reading all data onto the scope for
+     * import processing, should the user request an import.
+     *
+     * @argument {File} file
+     *     The file to upload onto the scope for further processing.
+     */
+    const handleFile = file => {
+
+        // The MIME type of the provided file
+        const mimeType = file.type;
+
+        // Check if the mimetype ends with one of the supported types,
+        // e.g. "application/json" or "text/csv"
+        if (_.every(LEGAL_FILE_TYPES.map(
+                type => !mimeType.endsWith(type)))) {
+
+            // If the provided file is not one of the supported types,
+            // display an error and abort processing
+            handleError(new ParseError({
+                message: "Invalid file type: " + mimeType,
+                key: 'IMPORT.ERROR_INVALID_FILE_TYPE',
+                variables: { TYPE: mimeType }
+            }));
+            return;
+        }
+
+        $scope.fileName = file.name;
+
+        // Initialize upload state
+        $scope.aborted = false;
+        $scope.dataReady = false;
+        $scope.processing = false;
+        $scope.uploadStarted = true;
+
+        // Save the MIME type to the scope
+        $scope.mimeType = file.type;
+
+        // Save the file to the scope when ready
+        $scope.fileReader = new FileReader();
+        $scope.fileReader.onloadend = (e => {
+
+            // If the upload was explicitly aborted, clear any upload state and
+            // do not process the data
+            if ($scope.aborted)
+                resetUploadState();
+
+            else {
+
+                // Save the uploaded data
+                $scope.fileData = e.target.result;
+
+                // Mark the data as ready
+                $scope.dataReady = true;
+
+                // Clear the file reader from the scope now that this file is
+                // fully uploaded
+                $scope.fileReader = null;
+
+            }
+        });
+
+        // Read all the data into memory
+        $scope.fileReader.readAsBinaryString(file);
+    };
+
+    /**
+     * Whether a drag/drop operation is currently in progress (the user has
+     * dragged a file over the Guacamole connection but has not yet
+     * dropped it).
+     *
+     * @type boolean
+     */
+    $scope.dropPending = false;
+
+    /**
+     * The name of the file that's currently being uploaded, or has yet to
+     * be imported, if any.
+     */
+    $scope.fileName = null;
+
+    /**
+     * The container for the file upload UI.
+     *
+     * @type Element
+     *
+     */
+    const uploadContainer = angular.element(
+            $document.find('.file-upload-container'));
+
+    /**
+     * The location where files can be dragged-and-dropped to.
+     *
+     * @type Element
+     */
+    const dropTarget = uploadContainer.find('.drop-target');
```

Comment:
Should this be moved to a directive such that the controller for this page does not directly manipulate the DOM?

Reply:
That's the way I originally had it, but the directive and controller got so intertangled that it ended up being cleaner to merge them.